src/date/converter.rs

```diff
@@ -1,4 +1,223 @@
-// TODO
-pub fn convert_gregorian_to_jalali(_year: u32, _month: u32, _day: u32) {}
-// TODO
-pub fn convert_jalali_to_gregorian(_year: u32, _month: u32, _day: u32) {}
+static DAY_SUM: [u32; 12] = [0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334];
+static DAY_SUM_KABISE: [u32; 12] = [0, 31, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335];
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct JalaliDate {
+    pub day: u32,
+    pub year: u32,
+    pub month: u32,
+}
+
+pub fn convert_gregorian_to_jalali(
+    year: u32,
+    month: u32,
+    day: u32,
+) -> Result<JalaliDate, &'static str> {
+    if month > 12 || month < 1 {
+        return Err("cant do it");
+    }
+
+    let days_sum = if is_kabise(year) {
+        DAY_SUM_KABISE[month as usize - 1] + day
+    } else {
+        DAY_SUM[month as usize - 1] + day
+    };
+
+    if days_sum < 79 {
+        let days_sum = days_sum + dey_jan_diff(year);
+        let jalai_year = year - 622;
+
+        if days_sum % 30 == 0 {
+            return Ok(JalaliDate {
+                year: jalai_year,
+                day: 30,
+                month: (days_sum / 30) + 9,
+            });
+        } else {
+            return Ok(JalaliDate {
+                year: jalai_year,
+                day: days_sum % 30,
+                month: (days_sum / 30) + 10,
+            });
+        }
+    } else {
+        let days_sum = days_sum - 79;
+        let jalali_year = year - 621;
+
+        if days_sum <= 186 {
+            if days_sum % 31 == 0 {
+                return Ok(JalaliDate {
+                    day: 31,
+                    year: jalali_year,
+                    month: days_sum / 31,
+                });
+            } else {
+                return Ok(JalaliDate {
+                    day: days_sum % 31,
+                    year: jalali_year,
+                    month: (days_sum / 31) + 1,
+                });
+            }
+        } else {
+            let days_sum = days_sum - 186;
+            if days_sum % 30 == 0 {
+                return Ok(JalaliDate {
+                    day: 30,
+                    year: jalali_year,
+                    month: (days_sum / 30) + 6,
+                });
+            } else {
+                return Ok(JalaliDate {
+                    day: days_sum % 30,
+                    year: jalali_year,
+                    month: (days_sum / 30) + 7,
+                });
+            }
+        }
+    }
+}
+
+fn dey_jan_diff(year: u32) -> u32 {
```

Comment:
Please choose a more descriptive name. And if possible, add a comment above it and describe it 🙏🏼

Reply:
Oh, dey is the month dey?

So, maybe something like this is more meaningful: `Gets the day difference between Persian month, Dey and Gregorian month January`